SQL: Resolve attributes recursively for improved subquery support

palesz opened the pull request:
Previously we did not resolve the attributes recursively which meant that if a field or expression was re-aliased multiple times (through multiple levels of subqueries), the aliases were only resolved one level down. This led to failed query translation because `ReferenceAttribute`s were pointing to non-existing attributes during query translation.

For example the query

```sql
SELECT i AS j FROM ( SELECT int AS i FROM test) ORDER BY j
```

failed during translation because the `OrderBy` resolved the `j` ReferenceAttribute to another `i` ReferenceAttribute that was later removed by an Optimization:

```
OrderBy[[Order[j{r}#4,ASC,LAST]]]                                             ! OrderBy[[Order[i{r}#2,ASC,LAST]]]
\_Project[[j]]                                                                = \_Project[[j]]
  \_Project[[i]]                                                              !   \_EsRelation[test][date{f}#6, some{f}#7, some.string{f}#8, some.string..]
    \_EsRelation[test][date{f}#6, some{f}#7, some.string{f}#8, some.string..] ! 
```

By resolving the `Attributes` recursively both `j{r}` and `i{r}` will resolve to `test.int{f}` above:

```
OrderBy[[Order[test.int{f}#22,ASC,LAST]]]                                     = OrderBy[[Order[test.int{f}#22,ASC,LAST]]]
\_Project[[j]]                                                                = \_Project[[j]]
  \_Project[[i]]                                                              !   \_EsRelation[test][date{f}#6, some{f}#7, some.string{f}#8, some.string..]
    \_EsRelation[test][date{f}#6, some{f}#7, some.string{f}#8, some.string..] ! 
 ```

The scope of recursive resolution depends on how the `AttributeMap` is constructed and populated.

Fixes #67237

Commit:
Resolve references recursively
Code simplification in the AttributeMap

elasticmachine commented:
Pinging @elastic/es-ql (Team:QL)

Commit:
Merge remote-tracking branch 'origin/master' into subquery-npe-fix

Commit:
Verifier error message fix

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/AttributeMap.java:
```diff
@@ -238,30 +238,52 @@ public boolean isEmpty() {
+     */
     @Override
     public E get(Object key) {
-        if (key instanceof NamedExpression) {
-            return delegate.get(new AttributeWrapper(((NamedExpression) key).toAttribute()));
-        }
-        return null;
+        return getOrDefault(key, null);
     }
 
+    /**
+     * @param key {@link NamedExpression} to look up
+     * @return Looks up the key in the AttributeMap recursively, meaning if the value for this key
+     * is another key in the map, it will follow it. It will return the final value or the
+     * specified default value if the key does not exist in the map.
+     */
     @Override
     public E getOrDefault(Object key, E defaultValue) {
-        E e;
-        return (((e = get(key)) != null) || containsKey(key))
-            ? e
-            : defaultValue;
+        E candidate = defaultValue;
+        E value = null;
+        while (((value = lookup(key)) != null || containsKey(key)) && value != key) {
+            key = candidate = value;
+        }
+        return candidate;
+    }
```
This is backwards.
getOrDefault should delegate to get not vice-versa. The former is an extension of the latter historically and conceptually.

Moreover by making get and getOrDefault recursive the Map contract has been fundamentally broken:
```
m.put(a, b);
assertTrue(b, m.get(a));
m.put(b,c);
assertTrue(b, m.get(a)); // fails
```

Instead of renaming get to `lookup` and modifying the contract, simply introduce a separate method (e.g. `traverse(key)`, `getResolvedValue(key)`, `getRecursive(key)` or something else) to the original map that does the recursive lookup and use that instead.

matriv replied:
I think I agree as well on this, I'd prefer to have a dedicated method that does this more complex logic rather than overriding the default `get`.

palesz replied:
Good points, but maybe the solution here is not just a simple rename of the method (`resolve(key)`), but to not implement the `Map<Attribute, E>` interface anymore. The Map contract was broken since the beginning because instead of `equals()` the `semanticEquals()` is used for key equality checks (which for example violates "containsKey(key) returns true if and only if this map contains a mapping for a key k such that (key==null ? k==null : key.equals(k)).")
After no`get()`s are required anymore, unless you can think of a place where recursive key resolution would actually cause harm (if that is the case, I'd say the AttributeMap building should be fixed instead to limit the scope from where the attributes are picked up from).

costin replied:
`AttributeMap` is a `Map` for `AttributeWrapper`s; the `get`, `contains` rely on `equals`. For convenience the underlying class is not exposed instead `Attribute` is used with the wrapping happening automatically.
It's similar to `IdentityHashMap` for example.

`resolve(key)` is a compound method that relies on `get`, the latter might not be used publicly now however I see too many downsides in removing it (and the `Map` contract) instead of keeping it.
For example `get`/`put` do per key read/write but `resolve`, which does multi-key read has not write counter-party; removing/replacing `get` breaks too many assumptions.

palesz replied:
I am not sure about the downsides or the breaking assumptions, since we barely utilize anything from the Map interface, practically only some read methods: `isEmpty`, `size` and `get/getOrDefault` (which can be renamed to `resolve`). The building of the map (`put`s) happen through the `Builder` methods today, here is how a full minimalization of the public methods would look like (at this point the `AttributeMap` is rather an `AttributeResolver`): https://github.com/palesz/elasticsearch/commit/fe814ec2e0c2a2aa4d082036603c79209c11be6a

Nonetheless, I can keep the current `Map` implementation and create a new `resolve` method for the recursive resolution, switch out all the `get/getOrDefault` calls to `resolve` and leave the `get`s there between the other unused methods.

palesz replied:
Done

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Verifier.java:
```diff
@@ -316,10 +316,11 @@ private static boolean checkGroupByOrder(LogicalPlan p, Set<Failure> localFailur
                 o.order().forEach(oe -> {
-                    Expression e = oe.child();
+                    final Expression e = oe.child();
+                    final Expression resolvedE = attributeRefs.getOrDefault(e, e);
```
See my comment above - it's misleading for a map lookup to do resolution.

palesz replied:
Done

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java:
```diff
@@ -131,8 +131,8 @@ public LogicalPlan optimize(LogicalPlan verified) {
 
         Batch refs = new Batch("Replace References", Limiter.ONCE,
-                new ReplaceReferenceAttributeWithSource()
-                );
+            new ReplaceReferenceAttributeWithSource()
+        );
 
         Batch operators = new Batch("Operator Optimization",
                 // combining
```
Noise

palesz replied:
Done

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java:
```diff
@@ -222,7 +222,7 @@ public LogicalPlan apply(LogicalPlan plan) {
             // collect aliases
             plan.forEachExpressionUp(Alias.class, a -> builder.put(a.toAttribute(), a.child()));
-            final Map<Attribute, Expression> collectRefs = builder.build();
+            final AttributeMap<Expression> collectRefs = builder.build();
```
Noise.

palesz replied:
Done

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryFolder.java:
```diff
@@ -600,10 +600,12 @@ else if (target.foldable()) {
                     GroupByKey matchingGroup = null;
                     if (groupingContext != null) {
-                        matchingGroup = groupingContext.groupFor(target);
+                        final Expression resolvedTarget = queryC.aliases().getOrDefault(target, target);
```
This bit of code is confusing due to the left overs.
Target doesn't seem to be used anywhere or at least its initial value is no good - instead of adding a `resolvedTarget` just reinitialize it:
`target = queryC.aliases().getOrDefault(target, target)`

Same for `id`.

palesz replied:
Done

costin commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java:
```diff
@@ -2658,4 +2657,29 @@ public void testSubqueryWithAliasOrderByAlias() throws Exception {
             "ORDER BY s.i > 10");
     }
+    
+    public void testReferenceResolutionInSubqueries() {
```
Please break this method into multiple tests that indicate the difference between them in their name.

palesz replied:
Done

costin reviewed (changes requested):
Glad to see progress on this however the breakage to the map contract isn't the way forward (see my comments).

Commit:
PR suggestions

Recursive get/getDefault() to resolve() in AttributeMap
Removed noise
Test breakdown

Commit:
Merge remote-tracking branch 'origin/master' into subquery-npe-fix

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/AttributeMap.java:
```diff
@@ -258,10 +260,21 @@ public E get(Object key) {
+        return defaultValue;
+    }
+
+    @SuppressWarnings({ "unchecked" })
+    public E resolve(Object key, E defaultValue) {
+        AttributeMap<Object> map = (AttributeMap<Object>) this;
+        Object candidate = defaultValue;
+        Object value = null;
+        while ((value = map.getOrDefault(key, NOT_FOUND)) != NOT_FOUND && value != key) {
+            key = candidate = value;
+        }
+        return (E) candidate;
     }
```
As this method builds on other types methods in this class, it shouldn't have unchecked warnings nor used `NOT_FOUND`. Also not sure why the `candidate` is returned instead of `value`, should be the other way around.

Further more this implementation looks buggy: if the map contains `(e,e)` calling resolve(e, default) would incorrectly return `default` and not `e`.
Further more there's no handling of cycles which leads to an infinite loop: (a, b), (b, a).

Below a suggestion to fix both issues:

```suggestion
    public E resolveOrDefault(Object key, E defaultValue) {
        E value = defaultValue;
        E candidate;
        int allowedLookups = 10;
        while (key != value && ((candidate = get(key)) != null || containsKey(key))) {
            if (--allowedLookups == 0) {
	           throw new QlIllegalArgumentException("Potential cycle detected");
	        }

            key = candidate;
            value = candidate;
        }
        return value;
    }
```

palesz replied:
Good catch with the default value. I was thinking of adding the limitation on the number of lookups, but seemed we would need quite a huge bug in our AttributeMap population for it to happen with more than one hops in the cycle. Anyways, I added this defensive check.

The suggested code won't really work, because:

1.`resolveOrDefault(e, e)` will always return `e` even if the map has `e -> f` mapping.
2. The loop will always terminate after the first lookup (`key == value == candidate`)

Anyways, the updated code fixes the bug you found.

costin commented on x-pack/plugin/ql/src/test/java/org/elasticsearch/xpack/ql/expression/AttributeMapTests.java:
```diff
@@ -61,6 +63,30 @@ public void testAttributeMapWithSameAliasesCanResolveAttributes() {
         assertTrue(newAttributeMap.get(param2.toAttribute()) == param2.child());
     }
 
+    public void testResolve() {
```
I'd like to see more tests for dealing with:
a. same key, value association - (e,e)
b. handling of default value across association: resolveOrDefault(k, default) for a map with (k, default), (default, value) and also `resolveOrDefault(k, k)`
c. handling of cycles: (a, b), (b, c), (c,a ) with a mixture of default value in place.

palesz replied:
✓ Done.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Verifier.java:
```diff
@@ -406,7 +411,7 @@ private static boolean checkGroupByHavingHasOnlyAggs(Expression e, Set<Expressio
         // resolve FunctionAttribute to backing functions
         if (e instanceof ReferenceAttribute) {
-            e = attributeRefs.get(e);
+            e = attributeRefs.resolve(e, null);
```
The null variant is used enough times to make a `resolve` method, in addition to `resolveOrDefault` on the main class similar to `get` and `getOrDefault` (even if it's just syntactic sugar).

palesz replied:
✓ Done.

Commit:
Bugfix and PR suggestions

Commit:
Merge remote-tracking branch 'origin/master' into subquery-npe-fix

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/AttributeMap.java:
```diff
@@ -266,15 +268,25 @@ public E getOrDefault(Object key, E defaultValue) {
+            if (--allowedLookups == 0) {
+                throw new QlIllegalArgumentException("Potential cycle detected");
+            }
+            key = value = candidate;
```
Please break the assignment in two pieces.

costin reviewed (approved):
LGTM

Commit:
PR suggestions

Commit:
Merge remote-tracking branch 'origin/master' into subquery-npe-fix
